modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,202 @@
+    _sliderCounter = 0;
+
+    // Parameters for learning
+    _lambda = lambda;  // regularization
+    _learnRate = learnRate;  // learning rate (ideally should be large at beginning and decay each iteration)
+    _nIterations = nIterations;  // number of training iterations
+
+    // True only in the first predict iteration
+    _initPredict = true;
+
+    // Online update flag
+    _onlineUpdate = true;
+
+    // Learn rate decay: _learnRate = _learnRate * _learnDecay
+    _learnRateDecay = 0.1;
+}
+
+SVMSGD::~SVMSGD(){
+
+}
+
+SVMSGD* SVMSGD::clone() const{
+    return new SVMSGD(*this);
+}
+
+void SVMSGD::train(cv::Mat trainFeatures, cv::Mat labels){
+
+    // Initialize _nFeatures
+    _slidingWindowSize = trainFeatures.rows;
+    _nFeatures = trainFeatures.cols;
+
+    float innerProduct;
+    // Initialize weights vector with zeros
+    if (_weights.size()==0){
+        _weights.reserve(_nFeatures);
+        for (uint feat = 0; feat < _nFeatures; ++feat){
+            _weights.push_back(0.0);
+        }
+    }
+
+    // Stochastic gradient descent SVM
+    for (uint iter = 0; iter < _nIterations; ++iter){
+        generateRandomIndex();
+        innerProduct = calcInnerProduct(trainFeatures.ptr<float>(_randomIndex));
+        int label = (labels.at<int>(_randomIndex,0) > 0) ? 1 : -1; // ensure that labels are -1 or 1
+        updateWeights(innerProduct, trainFeatures.ptr<float>(_randomIndex), label );
+    }
+}
+
+float SVMSGD::predict(cv::Mat newFeature){
+    float innerProduct;
+
+    if (_initPredict){
+        _nFeatures = newFeature.cols;
+        _slidingWindowSize = _predictSlidingWindowSize;
+        _featuresSlider = cv::Mat::zeros(_slidingWindowSize, _nFeatures, CV_32F);
+        _initPredict = false;
+        _labelSlider = new float[_predictSlidingWindowSize]();
+        _learnRate = _learnRate * _learnRateDecay;
+    }
+
+    innerProduct = calcInnerProduct(newFeature.ptr<float>(0));
+    
+    // Resultant label (-1 or 1)
+    int label = (innerProduct>=0) ? 1 : -1;
+
+    if (_onlineUpdate){
+        // Update the featuresSlider with newFeature and _labelSlider with label
+        newFeature.row(0).copyTo(_featuresSlider.row(_sliderCounter));
+        _labelSlider[_sliderCounter] = label;
+
+        // Update weights with a random index
+        if (_sliderCounter == _slidingWindowSize-1){
+            generateRandomIndex();
+            updateWeights(innerProduct, _featuresSlider.ptr<float>(_randomIndex), _labelSlider[_randomIndex]);
+        }
+
+        // _sliderCounter++ if < _slidingWindowSize
+        _sliderCounter = (_sliderCounter == _slidingWindowSize-1) ? 0 : (_sliderCounter+1);
+    }
+
+    return label;
+}
+
+void SVMSGD::generateRandomIndex(){
+    // Choose random sample, using Mikolov's fast almost-uniform random number
+    _randomNumber = _randomNumber * (unsigned long long) 25214903917 + 11;
+    _randomIndex = _randomNumber % (unsigned long long) _slidingWindowSize;
```

Comment:
_randomIndex is a 'unsigned int' , so too small for the assigned 'unsigned long long'